File: libs/async-core/src/com/satori/libs/async/core/AsyncCriticalSection.java
package com.satori.libs.async.core;

import com.satori.libs.async.api.*;

import java.util.*;
import java.util.function.*;

public class AsyncCriticalSection {
  
  private ArrayDeque<IAsyncHandler> awaiters = new ArrayDeque<>();
  private boolean entered = false;
  private boolean inloop = false;
  private final Consumer<Throwable> errorHandler;
  
  public AsyncCriticalSection() {
    this(null);
  }
  
  public AsyncCriticalSection(Consumer<Throwable> errorHandler) {
    this.errorHandler = errorHandler;
  }
  
  public void enter(IAsyncHandler cont) {
    if (entered) {
      awaiters.addLast(cont);
      return;
    }
    entered = true;
    try {
      cont.succeed();
    } catch (Throwable ex) {
      if (errorHandler != null) {
        errorHandler.accept(ex);
      }
    }
  }
  
  public IAsyncFuture enter() {
    if (entered) {
      AsyncFuture f = new AsyncFuture<>();
      awaiters.addLast(f);
      return f;
    }
    entered = true;
    return AsyncResults.succeeded();
  }
  
  public boolean tryEnter() {
    if (entered) {
      return false;
    }
    entered = true;
    return true;
  }
  
  public void leave() {
    if (!entered) {
      throw new IllegalStateException("trying to leave unentered critical section");
    }
    entered = false;
    if (inloop) {
      return;
    }
    inloop = true;
    do {
      IAsyncHandler cont = awaiters.pollFirst();
      if (cont == null) break;
      
      entered = true;
      try {
        cont.succeed();
      } catch (Throwable ex) {
        if (errorHandler != null) {
          errorHandler.accept(ex);
        }
      }
    } while (!entered);
    
    inloop = false;
  }
  
  public void exec(IAsyncHandler cont) {
    if (entered) {
      awaiters.addLast(ar -> {
        try {
          if (ar.isSucceeded()) cont.succeed();
          else cont.fail(ar.getError());
        } finally {
          leave();
        }
      });
      return;
    }
    entered = true;
    try {
      cont.succeed();
    } finally {
      leave();
    }
  }
}

File: libs/testlib/src/com/satori/libs/testlib/AsyncCriticalSection.kt
package com.satori.libs.testlib

import com.satori.libs.async.api.*
import com.satori.libs.async.core.*
import java.util.*

class AsyncCriticalSection {
  
  private val awaiters = ArrayDeque<IAsyncHandler<Unit>>()
  private var entered = false
  private var inloop = false
  
  fun enter(cont: IAsyncHandler<Unit>) {
    if (entered) {
      awaiters.addLast(cont)
      return
    }
    entered = true
    cont.succeed()
  }
  
  fun enter(): IAsyncFuture<Unit> {
    if (entered) {
      val f = AsyncFuture<Unit>()
      awaiters.addLast(f)
      return f
    }
    entered = true
    return AsyncResults.succeeded(Unit)
  }
  
  fun tryEnter(): Boolean {
    if (entered) {
      return false
    }
    entered = true
    return true
  }
  
  fun leave() {
    if (!entered) {
      throw IllegalStateException("trying to leave unentered critical section")
    }
    entered = false
    if (inloop) {
      return
    }
    inloop = true
    var error: Throwable? = null
    try {
      do {
        val cont = awaiters.pollFirst()
        if (cont === null) {
          return
        }
        entered = true
        try {
          cont.succeed()
        } catch (ex: Throwable) {
          error = ex // TODO: error can be overwritten
        }
      } while (!entered)
    } finally {
      inloop = false
      if (error != null) {
        throw error
      }
    }
  }
  
  fun exec(cont: IAsyncHandler<Unit>) {
    if (entered) {
      awaiters.addLast(IAsyncHandler { _ ->
        try {
          cont.succeed(Unit)
        } finally {
          leave()
        }
      })
      return
    }
    entered = true
    try {
      cont.succeed()
    } finally {
      leave()
    }
  }
}
